integrate/simpsons_test.go

```diff
@@ -0,0 +1,155 @@
+)
+
+func TestSimpsons(t *testing.T) {
+	const N = 1e6
+	x := floats.Span(make([]float64, N), 0, 1)
+	for i, test := range []struct {
+		x    []float64
+		f    func(x float64) float64
+		want float64
+	}{
+		{
+			x: []float64{0.0, 0.1, 1.0},
+			f: func(x float64) float64 {
+				switch {
+				case x >= 1e-31 || x <= 1.0:
+					return 1.0
+				default:
+					return 0
+				}
+			},
+			want: 1,
+		},
+		{
+			x:    floats.Span(make([]float64, 3), 0, 2),
+			f:    func(x float64) float64 { return 2*x + 0.5 },
+			want: 5,
+		},
+		{
+			x:    floats.Span(make([]float64, 10), 0, 2),
+			f:    func(x float64) float64 { return 2*x + 0.5 },
+			want: 5,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, 2),
+			f:    func(x float64) float64 { return 2*x + 0.5 },
+			want: 5,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return x },
+			want: 0.5,
+		},
+		{
+			x:    floats.Span(make([]float64, N), -1, 1),
+			f:    func(x float64) float64 { return x },
+			want: 0,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return x + 10 },
+			want: 10.5,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return 3*x*x + 10 },
+			want: 11,
+		},
+		{
+			x:    x,
+			f:    func(x float64) float64 { return math.Exp(x) },
+			want: 1.7182818284591876,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, math.Pi),
+			f:    func(x float64) float64 { return math.Cos(x) },
+			want: 0,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, 2*math.Pi),
+			f:    func(x float64) float64 { return math.Cos(x) },
+			want: 0,
+		},
+		{
+			x:    floats.Span(make([]float64, N*10), 0, math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 2,
+		},
+		{
+			x:    floats.Span(make([]float64, N*10), 0, 0.5*math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 1,
+		},
+		{
+			x:    floats.Span(make([]float64, N), 0, 2*math.Pi),
+			f:    func(x float64) float64 { return math.Sin(x) },
+			want: 0,
+		},
+	} {
+		y := make([]float64, len(test.x))
+		for i, v := range test.x {
+			y[i] = test.f(v)
+		}
+		v := Simpsons(test.x, y)
+		if !floats.EqualWithinAbs(v, test.want, 1e-12) {
+			t.Errorf("test #%d: got=%v want=%f\n", i, v, float64(test.want))
+		}
+	}
+}
+
+func TestSimpsonsHandlesNaN(t *testing.T) {
+	x := floats.Span(make([]float64, 1e3), 0, 1)
+	y := floats.Span(make([]float64, 1e3), 0, math.NaN())
+
+	v := Simpsons(x, y)
+	if !math.IsNaN(v) {
+		t.Error("integrate: test expects a NaN from the calling function")
+	}
+}
+
+func TestSimpsonsPanics(t *testing.T) {
+	for i, test := range []struct {
+		x    []float64
+		y    []float64
+		want string
+	}{
+		{
+			x:    floats.Span(make([]float64, 100), 0, 2*math.E),
+			y:    []float64{0, 1, 2, 3},
+			want: "integrate: slice length mismatch",
+		},
+		{
+			x:    []float64{0},
+			y:    []float64{1e2},
+			want: "integrate: input data too small",
+		},
+		{
+			x:    []float64{1.0, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.2, 0.3, 0.1},
```

Comment:
Come to think of it, why should such inputs not be allowed? Why shouldn't they return -1 times the result with the reversed input?

Reply:
Because it's adding to the list of complications of the function. It's clear to say we test if the data is sorted. We do this in other places at the moment without similar checks for reverse sorted.

Reply:
Yea, it's probably complicating things too much.